sigpy/mri/rf/slr.py

```diff
@@ -0,0 +1,648 @@
+    afa = mag2mp(np.sqrt(1 - np.abs(bf) ** 2))
+    a = np.fft.fft(afa) / Npad
+    a = a[0:N:1]
+    a = a[::-1]
+
+    return a
+
+
+def mag2mp(x):
+    N = np.size(x)
+    xl = np.log(np.abs(x))  # Log of mag spectrum
+    xlf = np.fft.fft(xl)
+    xlfp = xlf
+    xlfp[0] = xlf[0]  # Keep DC the same
+    xlfp[1:(N // 2):1] = 2 * xlf[1:(N // 2):1]  # Double positive frequencies
+    xlfp[N // 2] = xlf[N // 2]  # keep half Nyquist the same
+    xlfp[N // 2 + 1:N:1] = 0  # zero negative frequencies
+    xlaf = np.fft.ifft(xlfp)
+    a = np.exp(xlaf)  # complex exponentiation
+
+    return a
+
+
+def ab2rf(a, b):
+    N = np.size(a)
+    rf = np.zeros(N, dtype=complex)
+
+    a = a.astype(complex)
+    b = b.astype(complex)
+
+    for ii in range(N - 1, -1, -1):
+
+        Cj = np.sqrt(1 / (1 + np.abs(b[ii] / a[ii]) ** 2))
+        Sj = np.conj(Cj * b[ii] / a[ii])
+        theta = np.arctan2(np.abs(Sj), Cj)
+        psi = np.angle(Sj)
+        rf[ii] = 2 * theta * np.exp(1j * psi)
+
+        # remove this rotation from polynomials
+        if ii > 0:
+            at = Cj * a + Sj * b
+            bt = -np.conj(Sj) * a + Cj * b
+            a = at[1:ii + 1:1]
+            b = bt[0:ii:1]
+
+    return rf
+
+
+def rootFlip(b, d1, flip, tb):
+    # exhaustive root-flip pattern search for min-peak b1
+
+    n = np.size(b)
+    [w, bResp] = signal.freqz(b)
+    b /= np.max(np.abs(bResp))  # normalize beta
+    b *= np.sin(flip / 2 + np.arctan(d1 * 2) / 2)  # scale to target flip
+    r = sp.util.leja_fast(np.roots(b))
+
+    candidates = np.logical_and(np.abs(1 - np.abs(r)) > 0.004,
+                                np.abs(np.angle(r)) < tb / n * np.pi)
+
+    iiMin = 0
+    iiMax = 2 ** np.sum(candidates)
+
+    maxRF = np.max(np.abs(b2rf(b)))
+
+    for ii in range(iiMin, iiMax):
+
+        if ii % 20 == 0:
+            print('Evaluating root-flip pattern ' + str(ii) + ' out of ' +
+                  str(iiMax - iiMin))
+
+        # get a binary flipping pattern
+        doFlipStr = format(ii, 'b')
+        doFlip = np.zeros(np.sum(candidates), dtype=bool)
+        for jj in range(0, len(doFlipStr)):
+            doFlip[jj] = bool(int(doFlipStr[jj]))
+
+        # embed the pattern in an all-roots vector
+        tmp = np.zeros(n - 1, dtype=bool)
+        tmp[candidates] = doFlip
+        doFlip = tmp
+
+        # flip those indices
+        rFlip = np.zeros(np.shape(r), dtype=complex)
+        rFlip[:] = r[:]
+        rFlip[doFlip] = np.conj(1 / rFlip[doFlip])
+
+        bTmp = np.poly(rFlip)
+        [w, bTmpResp] = signal.freqz(bTmp)
+        bTmp /= np.max(np.abs(bTmpResp))  # normalize beta
+        bTmp *= np.sin(flip / 2 + np.arctan(d1 * 2) / 2)  # scale to targ. flip
+        rfTmp = b2rf(bTmp)
+
+        if np.max(np.abs(rfTmp)) < maxRF:
+            maxRF = np.max(np.abs(rfTmp))
+            rf_out = rfTmp
+            b_out = bTmp
+
+    return rf_out, b_out
+
+
+def dzRecursiveRF(Nseg, tb, N, seSeq=False, tbRef=8, zPadFact=4,
```

Comment:
Please change `dzRecursiveRF` to `dz_recursive_rf` and provide some documentation.